### ISSUE TITLE
Feature request: support MatchValid(PositionalArgs) OR MinimumValidNArgs, MaximumValidNArgs, RangeValidArgs

### ISSUE THREAD

**Issue by eine** [spf13/cobra#838]
For consistency, I think it would be useful to have `MinimumValidNArgs(int)`, `MaximumValidNArgs(int)` and `RangeValidArgs(min, max)` added to the list of supported `Args` formats. The complete list would then be:

- NoArgs
- ArbitraryArgs
- OnlyValidArgs
- MinimumNArgs(int)
- +**MinimumValidNArgs(int)**
- MaximumNArgs(int)
- +**MaximumValidNArgs(int)**
- ExactArgs(int)
- ExactValidArgs(int)
- RangeArgs(min, max)
- +**RangeValidArgs(min, max)**

This is because AFAIK it is not possible ATM to combine e.g. `MinimumNArgs(1)` with `OnlyValidArgs`.

---

An alternative solution is to provide a built-in function such as `matchAll` by @Antolius in [#745](https://github.com/spf13/cobra/issues/745#issuecomment-441326468). For example:

``` go
func MatchValid(p cobra.PositionalArgs) cobra.PositionalArgs {
	return func(cmd *cobra.Command, args []string) error {
		if err := cobra.OnlyValidArgs(cmd, args); err != nil {
			return err
		}
		if err := p(cmd, args); err != nil {
			return err
		}
		return nil
	}
}
```

which would allow the following expressions:

- `Args: cobra.MatchValid(cobra.MinimumNArgs(1)), ValidArgs: []string{...}`
- `Args: cobra.MatchValid(cobra.MaximumNArgs(5)), ValidArgs: []string{...}`
- `Args: cobra.MatchValid(cobra.RangeArgs(1, 5)), ValidArgs: []string{...}`
- `Args: cobra.MatchValid(cobra.ArbitraryArgs), ValidArgs: []string{...}`
- `Args: cobra.MatchValid(cobra.ExactArgs(1)), ValidArgs: []string{...}`

Certainly, `MatchValid` can be used implicitly when `ValidArgs` is defined, so the user is not required to type it. Therefore, `OnlyValidArgs` and `ExactValidArgs` can be deprecated from the public API. The list of formats would be:

- NoArgs
- ArbitraryArgs
- MinimumNArgs(int)
- MaximumNArgs(int)
- ExactArgs(int)
- RangeArgs(min, max)

**Comment by eine** [comment]
This is still an issue. It will be closed when #841 (or some other solution) is merged.

**Comment by johnSchnake** [comment]
Took the stale label away since the PR referenced that would fix this is still active.

**Comment by marckhouzam** [comment]
The alternative solution proposed in the description of this issue is now available thanks to #896 and the new `MatchAll(pargs ...PositionalArgs)` function: https://github.com/spf13/cobra/blob/03c3eb73f8d00531742e232eb311f8cce0cd6225/args.go#L111-L112

I'm going to go ahead and close this as fixed.
